src/decoder/cuda-decoder-utils.cu

```diff
@@ -0,0 +1,237 @@
+// decoder/cuda-decoder-utils.cc
+
+// Copyright      2018  Zhehuai Chen
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http:// www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "fst/fstlib.h"
+#include "cuda-decoder-utils.h"
+
+namespace kaldi {
+
+void get_free_memory_stat(const char *prefix) {
+#if 0 // TODO: define of GetFreeMemory
+  int32 act_gpu_id;
+  cudaError_t e = cudaGetDevice(&act_gpu_id);
+  char name[128];
+  CuDevice::Instantiate().DeviceGetName(name, 128, act_gpu_id);
+  // get GPU memory stats
+  int64 free, total;
+  std::string mem_stats;
+  mem_stats = CuDevice::Instantiate().GetFreeMemory(&free, &total);
+  float mem_ratio = total > 0 ? free / (float)total : 0;
+  // log
+  KALDI_LOG << prefix << "\tcudaSetDevice(" << act_gpu_id << "): "
+            << name << "\t" << mem_ratio << " % "
+            << "\t" << mem_stats;
+#endif
+}
+
+// a combination of cudaMallocManaged & cudaMemAdvise
+void cuda_malloc_managed_preferred_device(void** devPtr, size_t size) {
+  cudaMallocManaged(devPtr, size);
+#ifdef MEMADVISE
```

Comment:
doing this does give a performance improvement. using ifdef here is mainly because:

#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 600
#define MEMADVISE // used after Pascal, details in:
// http:// mug.mvapich.cse.ohio-state.edu/static/media/mug/presentations/2016/MUG16_GPU_tutorial_V5.pdf
#endif